### ISSUE TITLE
Outdated Source HM Note 3 aka kenzo

### ISSUE THREAD

**Issue by khanfrd** [MiCode/Xiaomi_Kernel_OpenSource#60]
The **new Redmi Note 3** uses **goodix fingerprint drivers**, and the source is outdated
**so please update the source ASAP**

**Comment by sysxda** [comment]
+1

**Comment by alesasnouski** [comment]
Guys, can you update source code ?

**Comment by InfestAl** [comment]
Confirm, please update the source

**Comment by happybydefault** [comment]
Update the source code.

**Comment by arhue** [comment]
This needs urgent attention. Changing hardware has broken certain things on AOSP based ROMs. I think some people are reporting VoLTE and front camera not working?

**Comment by roastedwildebeest** [comment]
Please update the kernel source. Fingerprint and Front Camera are not working.

**Comment by Tyaginator** [comment]
Update update update update update!

**Comment by Fosnine** [comment]
Hi guys, updating the source code would be great.
Since new type of fingerprint sensor and front camera are not working in AOSP based ROM. :)

**Comment by miji2** [comment]
+1

**Comment by acnalogia** [comment]
Please update the kernel source.
fingerprint sensor and front camera are not working in AOSP based ROM

**Comment by Maria-gr** [comment]
Update the sources please !! Fingerprint and front camera not working on newer models..

**Comment by arhue** [comment]
Actually it's not that Xiaomi can choose to not update the sources. They are forced to since the kernel uses GPLv2 and they already have distributed the updated kernel to the users. I know copyright doesn't mean anything in China but they have released this product in India so they can be sued for this.

**Comment by gingerboy92** [comment]
Update the source code please!!!

**Comment by khanfrd** [comment]
It's been week no reply from xiaomi, that what I expected 
anyway I got in touch with yijiangtao author of our goodix driver, he say the the current code of goodix we have right now are fine they should work, the issue is with DTS or maybe hall,
I will take a look again, it will be nice if other dev take a look too

**Comment by alesasnouski** [comment]
What about front camera ?

**Comment by khanfrd** [comment]
I have the newer hardware one and don't have any issue with cam on miui

**Comment by yemjalakumar** [comment]
Update the source code please !

**Comment by Whitexp** [comment]
xiaomi , we waiting for u update the source .

**Comment by tank0412** [comment]
Hello, guys.
I am a rom developer. Kernel sources by Xiaomi has **goodix and front camera driver**.
**Again. Sources by Xiomi is COMPLETE.**
You can find goodix driver there:
https://github.com/MiCode/Xiaomi_Kernel_OpenSource/tree/kenzo-l-oss/drivers/input/misc/goodix
So you may ask me why it does not work on custom rom.
It does not work because **developer of official CM removed goodix driver from kernel sources.**
Also he removed all files for goodix and front camera in rom.
Check my commits:
fix of front camera:
https://github.com/tank0412/proprietary_vendor_xiaomi/commit/7301b1859bf07805e05a1d844bc6409e124f135a
experiments with goodix blobs;
https://github.com/tank0412/proprietary_vendor_xiaomi/commit/e90d0ea1aca52490eb8e472a14ccdc782921a57d
https://github.com/tank0412/proprietary_vendor_xiaomi/commit/454a8283b8a4a72ccf65fcd4e62e8318457cbbb9
https://github.com/tank0412/proprietary_vendor_xiaomi/commit/0a53dc3432f9bb58b6e0816c886b458fc1e26d10
So it is not a fault of Xiaomi. It is a fault of TheStrix.

P.S. I fixed front camera issue on officail CM builds and published fix. Are you still do not know about it???
http://forum.xda-developers.com/showpost.php?p=68183787&postcount=18

**Comment by alesasnouski** [comment]
Can you give us a link to your PR? Thank you.

**Comment by tank0412** [comment]
http://forum.xda-developers.com/showpost.php?p=68183787&postcount=18

**Comment by khanfrd** [comment]
@tank0412 first of all its no one fault not all the user have goodix fp,
and some user reported goodix fp doesn't work on miui too i don't know exactly as i have fpc1020

and yes i am too still confused with TheStrix, i dont know why he use MM blobs, as much i know about roms we can use lp blobs too

and as you say you are rom developer can you help me out with something???

and if any user with goodix can confirm and clean this misunderstanding about goodix fp on miui it will be great help
if goodix with custom kernel works on miui then there is no issue with source,
will close the issue them

**Comment by alex9yust** [comment]
@tank0412 how run gx_fpd? in miui sets fingerprintd and gx_fpd, and fingerprintd from cyanogenmod knows nothing about gx_fpd.
it certainly does not belong to the kernel.

**Comment by arhue** [comment]
"and if any user with goodix can confirm and clean this misunderstanding about goodix fp on miui it will be great help
if goodix with custom kernel works on miui then there is no issue with source,"
I think you should wait for some senior developer to confirm. Regardless if Xiaomi had updated the kernel during the last 5 months after they released the original sources, they have to update the sources to match the released kernel version technically. But I think we can give them a pass on it. Just wait for someone to confirm to sort the issue out.

**Comment by poad42** [comment]
@tank0412 Buddy if you are a ROM developer you'd know my taking a look at cm sources even the FPC driver used is not from xiaomi's sources but knaged from another device. To even make use of xiaomi's sources they'd have to open source the fingerprint api that they used in miui lollipop, which they won't obviously. So we cannot use the existing goodix and FPC drivers because they are all outdated. Regarding the camera sensor, xiaomi's has only sourced one of them while using two types of sensor for their units same goes to gyroscope as well. Therefore they have to update their and its not complete by a country mile.

**Comment by tank0412** [comment]
@khanfrd I have phone with goodix fingerprint. And it works on MIUI MM **sometimes** (but it works)
We can use lollipop blobs **BUT** we need libshims for it and AFAIK RIL does not work on Lollipop blobs. (correct if i wrong)
Even if use LP blobs, fingerprint will not work due to major differentces in fingerprint MM API.

**Comment by khanfrd** [comment]
@tank0412 what about lp?? goodix work with custom kernel on lp??

**Comment by GuneetAtwal** [comment]
goodix doesnt work because it is broken! i  and @nicknitewolf have tried almost every method to fix goodix fp and we have got to making it appear in the fp settings.

TheStrix removed it because they were not able to fix the issue. Santosh posted his goodix fix but never open sourced it due to which we have to face so many problems.

**Comment by tank0412** [comment]
@GuneetAtwal But Santosh M just port kernel from Miui MM to CM13. Nothing else.
So it can not be opensourced.

**Comment by GuneetAtwal** [comment]
nope, he made a patch for goodix sensor , also his device trees are closed sourced

**Comment by niclimcy** [comment]
@GuneetAtwal true. He does not open the source. Even Xiaomi updated their own internal sources of L yet they never release.

**Comment by arhue** [comment]
@GuneetAtwal Is it some kind of a shim that he has made for the FP API?
He is not forced to open source it  thanks to the permissive Apache license but we can pressure him to.

**Comment by GuneetAtwal** [comment]
why pressurise him ? let him do whatever he wants to do. If he doesn't care about us we also dgaf.

**Comment by securerootd** [comment]
And now Global Beta and China Beta for MM is out, and we do not have yet the kernel sources.
We need the sources to be updated to MM with all sensors, so custom rom/kernel support can complete

**Comment by N000bak** [comment]
Please update the kernel source. Fingerprint and Front Camera are not working.

**Comment by sumchamp** [comment]
Xiaomi, please also release MM sources for VoLTE.

**Comment by tmaone** [comment]
http://gpl-violations.org/

**Comment by khanfrd** [comment]
Now we know source is complete for Lollipop
And official Marshmallow Rom is yet to come, So we cant bug Xiaomi yet
So closing the issue

**Comment by LightTube** [comment]
MM realeased now, open it

**Comment by yemjalakumar** [comment]
Official global beta mm  released and it is stable too

**Comment by poad42** [comment]
Drop it guys, they won't give us an updated kernel source. Knowing xiaomi's history of GPL violations, they are never going to update the source.

**Comment by khanfrd** [comment]
As we know [official marshmallow](http://en.miui.com/download-301.html) is up and according to gpl kernel source need to update too
So Xiaomi please update fucking kernel source too!!!!!!!!!!!!

MIUI thread bug them here too guys 
http://en.miui.com/thread-254760-1-1.html

**Comment by naymuln** [comment]
+1

**Comment by AmolAmrit** [comment]
Even If many things got fixed now. We still have the old Kernel sources. So please Xiaomi release kernel sources.

**Comment by itsjoeoui** [comment]
+1

**Comment by itsjoeoui** [comment]
Plz update it
